File: app/src/main/java/com/projects/johnny/myway/SignInFragment.java
package com.projects.johnny.myway;

import android.app.Fragment;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.util.Patterns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.firebase.client.AuthData;
import com.firebase.client.Firebase;
import com.firebase.client.FirebaseError;

import java.util.regex.Pattern;

// TODO: Merge SignInFragment and SignUpFragment into one Activity

public class SignInFragment extends Fragment {

    private EditText mUsername;
    private EditText mPassword;
    private TextView mUsernameRetry;
    private TextView mSignUpButton;
    private Button mSignInButton;

    public static final String USER_UID_KEY = "USER_UID_KEY";

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_sign_in, container, false);

        // Get reference to Firebase
        // TODO: Move Firebase to parent activity
        Firebase.setAndroidContext(getActivity());
        final Firebase mFirebaseRef = new Firebase("https://myways.firebaseIO.com/");

        mUsername = (EditText) v.findViewById(R.id.sign_in_username_edit_text);
        mPassword = (EditText) v.findViewById(R.id.sign_in_password_edit_text);
        mUsernameRetry = (TextView) v.findViewById(R.id.username_retry_text_view);
        mSignUpButton = (Button) v.findViewById(R.id.sign_up_button);
        mSignInButton = (Button) v.findViewById(R.id.sign_in_button);

        mUsernameRetry.setVisibility(View.GONE);

        mUsername.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (!hasFocus) {
                    String username = mUsername.getText().toString();
                    if (isValidUsername(username)) {
                        mUsernameRetry.setVisibility(View.GONE);
                    } else {
                        mUsernameRetry.setVisibility(View.VISIBLE);
                    }
                }
            }
        });
        mSignUpButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // TODO: Automatically log user in after creating account
                Intent intent = new Intent(getActivity(), SignUpActivity.class);
                startActivity(intent);
            }
        });
        mSignInButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (isValid()) {
                    final String email = mUsername.getText().toString();
                    final String password = mPassword.getText().toString();


                    mFirebaseRef.authWithPassword(email, password, new Firebase.AuthResultHandler() {
                        @Override
                        public void onAuthenticated(AuthData authData) {
                            // Use App class for session storage of UID
                            String UID = authData.getUid();
                            App.Companion.setUID(UID);
                            Intent intent = new Intent(getActivity(), MainActivity.class);
                            startActivity(intent);
                            getActivity().finish();
                        }

                        @Override
                        public void onAuthenticationError(FirebaseError firebaseError) {
                            Toast.makeText(getActivity().getApplicationContext(), "Failed to log in...", Toast.LENGTH_SHORT).show();
                            System.out.println(firebaseError.getMessage());
                        }
                    });
                }
            }
        });

        return v;
    }

    // Returns true/false depending on whether locations is turned on or off.
    private static boolean isLocationEnabled(Context context) {
        int locationMode = 0;
        String locationProviders;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT){
            try {
                locationMode = Settings.Secure.getInt(context.getContentResolver(), Settings.Secure.LOCATION_MODE);

            } catch (Settings.SettingNotFoundException e) {
                e.printStackTrace();
            }

            return locationMode != Settings.Secure.LOCATION_MODE_OFF;

        } else{
            locationProviders = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.LOCATION_PROVIDERS_ALLOWED);
            return !TextUtils.isEmpty(locationProviders);
        }
    }

    private boolean isValidUsername(CharSequence username) {
        Pattern pattern = Patterns.EMAIL_ADDRESS;
        return pattern.matcher(username).matches();
    }

    private boolean isValid() {
        String username = mUsername.getText().toString();
        String password = mPassword.getText().toString();
        if (username.length() > 0 && password.length() > 0) {
            return true;
        }
        return false;
    }
}

File: app/src/main/java/com/projects/johnny/myway/SignInFragment.kt
package com.projects.johnny.myway

import android.app.Fragment
import android.os.Bundle
import android.util.Log
import android.util.Patterns
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.credentials.Credential
import com.google.android.gms.common.api.Status
import kotlinx.android.synthetic.main.fragment_sign_in.*
import org.jetbrains.anko.onClick

class SignInFragment : Fragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(R.layout.fragment_sign_in, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        usernameRetry.visibility = View.GONE

        username.setOnFocusChangeListener { view, hasFocus ->
            if (!hasFocus) {
                val name = username.text.toString()
                val visibility = if (isValidUsername(name)) View.GONE else View.VISIBLE
                usernameRetry.visibility = visibility
            }
        }
        signUpButton.onClick { // onClick belongs to anko
            // TODO: set up listener to start SignUpFragment
        }
        signInButton.setOnClickListener {
            if (isValid()) {
                with(activity as MainActivity) {
                    val email = username.text.toString()
                    val pass = password.text.toString()
                    val credential = Credential.Builder(email)
                                        .setName(email)
                                        .setPassword(pass)
                                        .build()
                    Auth.CredentialsApi.save(credentialsClient, credential).setResultCallback {
                        status: Status ->
                        Log.d("Credentials", status.toString())
                        if (status.isSuccess) {
                            Log.i("SignInFragment", "Credentials saved!")
                            authenticateWithCredential(credential)
                        } else {
                            if (status.hasResolution()) {
                                // Try to resolve the save request. This will prompt
                                // the user if the credential is new.
                                status.startResolutionForResult(activity, MainActivity.rc_save)
                                authenticateWithCredential(credential)
                            } else {
                                Log.e("CredentialsApi.save", "Save failed!")
                            }
                        }
                    }
                }
            }
        }
    }

    private fun isValidUsername(name: CharSequence): Boolean =
            Patterns.EMAIL_ADDRESS.matcher(name).matches()

    // Checks if username & password entries are valid in length
    private fun isValid(): Boolean {
        val name = username.text.toString()
        val pass = password.text.toString()
        return if (name.length > 0 && pass.length > 0) true else false
    }
}